src/fastfield/multivalued/writer.rs

```diff
@@ -200,24 +184,169 @@ impl MultiValuedFastFieldWriter {
-                let (val_min, val_max) = val_min_max.unwrap_or((0u64, 0u64));
-                value_serializer =
-                    serializer.new_u64_fast_field_with_idx(self.field, val_min, val_max, 1)?;
                 for vals in self.get_ordered_values(doc_id_map) {
                     // sort values in case of remapped doc_ids?
                     for &val in vals {
-                        value_serializer.add_val(val)?;
+                        values.push(val);
                     }
                 }
             }
-            value_serializer.close_field()?;
+            let col = VecColumn::from(&values[..]);
+            serializer.create_auto_detect_u64_fast_field_with_idx(self.field, col, 1)?;
         }
         Ok(())
     }
 }
+
+pub(crate) struct MultivalueStartIndex<'a, C: Column> {
+    column: &'a C,
+    doc_id_map: &'a DocIdMapping,
+    min_max_opt: Mutex<Option<(u64, u64)>>,
+    random_seeker: Mutex<MultivalueStartIndexRandomSeeker<'a, C>>,
+}
+
+struct MultivalueStartIndexRandomSeeker<'a, C: Column> {
+    seek_head: MultivalueStartIndexIter<'a, C>,
+    seek_next_id: u64,
+}
+
+impl<'a, C: Column> MultivalueStartIndex<'a, C> {
+    pub fn new(column: &'a C, doc_id_map: &'a DocIdMapping) -> Self {
+        assert_eq!(column.num_vals(), doc_id_map.num_old_doc_ids() as u64 + 1);
+        MultivalueStartIndex {
+            column,
+            doc_id_map,
+            min_max_opt: Mutex::default(),
+            random_seeker: Mutex::new(MultivalueStartIndexRandomSeeker {
+                seek_head: MultivalueStartIndexIter {
+                    column,
+                    doc_id_map,
+                    new_doc_id: 0,
+                    offset: 0u64,
+                },
+                seek_next_id: 0u64,
+            }),
+        }
+    }
+
+    fn minmax(&self) -> (u64, u64) {
+        if let Some((min, max)) = self.min_max_opt.lock().unwrap().clone() {
+            return (min, max);
+        }
+        let (min, max) = tantivy_bitpacker::minmax(self.iter()).unwrap_or((0u64, 0u64));
+        *self.min_max_opt.lock().unwrap() = Some((min, max));
+        (min, max)
+    }
+}
+impl<'a, C: Column> Column for MultivalueStartIndex<'a, C> {
+    fn get_val(&self, idx: u64) -> u64 {
+        let mut random_seeker_lock = self.random_seeker.lock().unwrap();
+        if random_seeker_lock.seek_next_id > idx {
+            *random_seeker_lock = MultivalueStartIndexRandomSeeker {
+                seek_head: MultivalueStartIndexIter {
+                    column: self.column,
+                    doc_id_map: self.doc_id_map,
+                    new_doc_id: 0,
+                    offset: 0u64,
+                },
+                seek_next_id: 0u64,
+            };
+        }
+        let to_skip = idx - random_seeker_lock.seek_next_id;
+        random_seeker_lock.seek_next_id = idx + 1;
+        random_seeker_lock.seek_head.nth(to_skip as usize).unwrap()
```

Comment:
I think assuming that get_val calls always come in order is too dangerous. Multiple codec estimations would already violate that assumption